Fix/cmake codegen

Ashutosh-Londhe opened the pull request.

reguly commented on ops_translator/CMakeLists.txt:
```diff
@@ -0,0 +1,44 @@
+    INSTALL(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}" DESTINATION bin)
+
+#   The following operations are OS dependent, we do UNIX first
+    if (UNIX)
```
any chance of doing this for windows too? We do have some users there

Ashutosh-Londhe replied:
I am not sure for windows, i need to check how this can be done in CMAKE

reguly commented on CMakeLists.txt:
```diff
@@ -1,253 +1,322 @@
+            set(CMAKE_CUDA_ARCHITECTURES
+                70
+                CACHE STRING "CUDA architectures")
+                message(WARNING "Using Volta architecure for compiling CUDA. \nIf different please provide correct GPU architecture using -DGPU_ARCH=XXX  where XXX is compute capability of an architecture!")
```
add: such as 70 if you want to target sm_70

reguly commented on CMakeLists.txt:
```diff
@@ -1,253 +1,322 @@
+#   option(BUILD_OPS_FORTRAN "Build FORTRAN version of OPS" OFF)
+    option(BUILD_OPS_APPS "Build Sample OPS applications" OFF)
+
+#   option(OPS_ERROR_ON_WARNING "Error upon encountering a warning" ON)
```
why leave these commented out ones?

Ashutosh-Londhe replied:
It is commented only, not sure if it was tested before

reguly reviewed (changes requested):
How am I supposed to start building with CMake? When I build with 
cmake .. -DBUILD_OPS_APPS=ON
I get
Code-gen command:  /usr/local/bin/ops_translator/ops-translator -DOPS_ACC_IGNORE -I /usr/local/include --file_paths
 PdV.cpp;accelerate.cpp;advec_cell.cpp;advec_mom.cpp;calc_dt.cpp;clover_leaf.cpp;field_summary.cpp;flux_calc.cpp;generate.cpp;ideal_gas.cpp;initialise_chunk.cpp;reset_field.cpp;revert.cpp;update_halo.cpp;viscosity.cpp
 CMake Error at apps/c/CMakeLists.txt:560 (message):
   OPS Python code generation command failed with error code No such file or
   directory:
 Call Stack (most recent call first):
   apps/c/CloverLeaf/CMakeLists.txt:12 (build_ops_c_sample)

Ashutosh-Londhe commented:
> How am I supposed to start building with CMake? When I build with cmake .. -DBUILD_OPS_APPS=ON I get Code-gen command: /usr/local/bin/ops_translator/ops-translator -DOPS_ACC_IGNORE -I /usr/local/include --file_paths PdV.cpp;accelerate.cpp;advec_cell.cpp;advec_mom.cpp;calc_dt.cpp;clover_leaf.cpp;field_summary.cpp;flux_calc.cpp;generate.cpp;ideal_gas.cpp;initialise_chunk.cpp;reset_field.cpp;revert.cpp;update_halo.cpp;viscosity.cpp CMake Error at apps/c/CMakeLists.txt:560 (message): OPS Python code generation command failed with error code No such file or directory: Call Stack (most recent call first): apps/c/CloverLeaf/CMakeLists.txt:12 (build_ops_c_sample)

Can you please check if ops_translator is available in the path you mentioned when installing OPS LIB. 

I have used following commands:

Building OPS lib

1. Loaded the required modules (module load cuda/toolkit-10.0.130 gnu-6.3.0/ompi-1.10.6 python/3.9.7 cmake)
2. Inside OPS directory created build folder
3. From build folder
    >   cmake -DCMAKE_INSTALL_PREFIX=<path where u want to install OPS and its dependencies> -DHDF5_ROOT=<path to HDF5 home> ..
    >   make
    >   make install   
    (make sure python dependencies are installed after this. make install will normally take longer as it will install the python packages required for code-gen to function)
 
 Building apps
 4. I do not tried building all apps, but able to build individually poisson as well as Cloverleaf
     for poisson and cloverleaf
     i.  copied CMakelist.txt from OPS/apps/c to required folder
     ii. deleted all add_directory present at end of the file and replaced with 
         for poisson: 
                            build_ops_c_sample(poisson "NONE" "NONE" "NONE" "NO" "YES")  
         for cloverleaf: 
                            set(others start.cpp time_step.cpp build_field.cpp read_input.cpp
                                    initialise.cpp advection.cpp)
                             build_ops_c_sample(clover_leaf profile "${others}" "NONE" "NO" "YES")
      iii. create build folder inside c/poisson or c/Cloverleaf and then run  
          >  cmake -DOPS_INSTALL_DIR=<path where OPS lib is installed with earlier cmake command> -DHDF5_ROOT=<hdf5 home> ..
          
I have used this with GNU compiles on Telos and able to generate and run the application 

If you are using Intel or PGI compiler, please add -DENABLE_IEEE=on to your cmake command so that IEEE flags will be added to CMAKE_C and CMAKE_CXX flags when compiling application which are sensitive to precision

reguly commented:
We need to properly support -DBUILD_OPS_APPS=ON  without all the extra steps. Or remove the option. But I would much prefer fixing it, because having all the extra steps sort of defeats the purpose of a cmake build (which is supposed to "just work")

reguly commented:
We should not try to build hdf5-based apps if we don't have HDF5 installed. Currently these are getting built, and fails when it comes to linking them.

Ashutosh-Londhe commented:
> We should not try to build hdf5-based apps if we don't have HDF5 installed. Currently these are getting built, and fails when it comes to linking them.

fixed

gihanmudalige commented:
Is this now ready to be merged ?
